lib/TH/THGeneral.c

```diff
@@ -137,10 +159,17 @@ void* THRealloc(void *ptr, long size)
   if(size < 0)
     THError("$ Torch: invalid memory size -- maybe an overflow?");
 
-  ptr = realloc(ptr, size);
-  if(!ptr)
+  void *newptr = realloc(ptr, size);
+
+  if(!newptr && torchGCFunction) {
+    torchGCFunction(torchGCData);
+    newptr = realloc(ptr, size);
+  }
+
+  if(!newptr)
     THError("$ Torch: not enough memory: you tried to reallocate %dGB. Buy new RAM!", size/1073741824);
```

Comment:
Calling `lua_error` doesn't mean your process will exit (e.g. if this code was executed within a `pcall`). So you have to consider freeing memory before calling THError to avoid leaks.